_rules/SVG-image-has-name-7d6734.md

```diff
@@ -0,0 +1,205 @@
+---
+id: 7d6734
+name: SVG image has accessible name
+rule_type: atomic
+description: |
+   This rule checks that each SVG image element that renders visible content that and is included in the accessibility tree has an accessible name.
+accessibility_requirements:
+  wcag20:1.1.1: # Non-Text Content (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+authors:
+  - Bryn Anderson
+---
+
+## Applicability
+
+The rule applies to any element in the [SVG](https://www.w3.org/2000/svg) namespace with a [semantic role](#semantic-role) of either `img`, `graphics-document`, `graphics-object`, `graphics-symbol`, that is [visible](#visible), and [included in the accessibility tree](#included-in-the-accessibility-tree).
+
+**Note**: The [SVG Accessibility API Mappings](https://www.w3.org/TR/svg-aam-1.0/#include_elements) specifies that many elements in the SVG namespace are purely presentational and should not be included in the accessibility tree unless indicated otherwise through the use of text alternative content, an explicit WAI ARIA role, or a valid `tabindex` attribute.
+
+## Expectation
+
+Each target element has an [accessible name](#accessible-name) that is not [whitespace](#whitespace).
+
+## Assumptions
+
+*There are currently no assumptions*
+
+## Accessibility Support
+
+The [HTML Accessibility API Mappings](https://www.w3.org/TR/html-aam-1.0/#html-element-role-mappings) specify that the `<svg>` element has an implicit role of `graphics-document`. However browser support for the `graphics-document` role and the [SVG Accessibility API Mappings](https://www.w3.org/TR/svg-aam-1.0) is inconsistent.
+
+Browser and assistive technology support for SVG `<title>` and `<desc>` elements is currently inconsistent. Using WAI ARIA in combination with the `img` role for non-decorative `<svg>` elements significantly improves accessibility browser support.
+
+## Background
+
+- [Understanding Success Criterion 1.1.1: Non-text Content](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html)
+- [SVG Accessibility API Mappings](https://www.w3.org/TR/svg-aam-1.0)
+- [WAI-ARIA Graphics Module](https://www.w3.org/TR/graphics-aria-1.0/)
+- [ARIA4: Using a WAI-ARIA role to expose the role of a user interface component](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA4)
+- [ARIA6: Using aria-label to provide labels for objects](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA6)
+- [ARIA10: Using aria-labelledby to provide a text alternative for non-text content](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA10)
+
+## Test Cases
+
+### Passed
```

Comment:
Apologies for not bringing this up sooner, but strictly speaking, none of the SVGs you have here convey information. That means they're purely decorative and so they should be marked up in a way that assistive technologies can ignore them. I suggest you find simple SVG shapes that do actually need an accessible name.

Reply:
@WilcoFiers how would you feel about an assumption that says something like, "for the purpose of demonstrating this rule the test cases that include content in the accessibility tree should be assumed to be conveying meaning"

I spoke to @Jym77 and we both understand where your coming from but feel that for the purpose of demonstration the test cases are perfectly acceptable. It is true that by default the <circle> element has no semantic meaning and does not convey any information, however the author can change this by either providing:
-  a title or desc element containing content,
- an aria-label or aria-labelledby attribute with a value,
- an aria-describedby attribute with a valid reference,
- a valid tabindex,
- an allowed explicit role other than none or presentation, or
- if it has any other characteristics noted in the Core AAM for being included in the acc tree.

I think we have three options:
either we add an assumption as suggested above, or
we build new test cases that convey meaning - e.g. a diagram like the one for an electrical switch below (seriously bulky), or
you provide a test case theme you would be happy with.

**Electrical circuit diagram**
```html
<svg xmlns="http://www.w3.org/2000/svg" width="400" height="200" viewBox="0 0 200 100" role="graphics-document document" xml:lang="en">
     <title>A simple circuit</title>
     <desc>A circuit with one source, one switch, and one load</desc>
     <style type="text/css">
        svg { /* default styles */
           fill: none;
           stroke: black;
        }
        .wire {
           stroke-width: 2px;
        }
        .connection {
           stroke: none;
           fill: black;
        }
        text {
           fill: black;
           stroke: none;
           font: 15px sans-serif;
        }
     </style>
     <g id="battery-1" role="graphics-symbol img" aria-roledescription="source" aria-label="battery" transform="translate(20,50)">
        <path d="M-15,-5 h30 M-5,5 h10"/>
     </g>
     <path id="wire-1" role="graphics-symbol img" aria-label="wire connecting" aria-labelledby="wire-1 battery-1 switch-1" class="wire" d="M20,45 V20 H90"/>
     <g id="switch-1" role="graphics-symbol img" aria-roledescription="switch" aria-label="single-pole switch">
        <path d="M90,20 L105,30"/>
        <circle class="connection" cx="90" cy="20" r="2"/>
        <circle class="connection" cx="110" cy="20" r="2"/>
     </g>
     <path id="wire-2" role="graphics-symbol img" aria-label="wire connecting" aria-labelledby="wire-2 switch-1 lightbulb-1" class="wire" d="M110,20 H180 V40"/>
     <g id="lightbulb-1" role="graphics-symbol img" aria-roledescription="load" aria-label="lightbulb" aria-labelledby="lightbulb-1 lightbulb-1-label" transform="translate(180,50)">
         <text id="lightbulb-1-label" x="-15" dy="0.5ex" text-anchor="end">10W</text>
         <circle r="10"/>
         <path d="M0,-10 C0,8 5,8 5,0 C5,-8 0,-8 0,10"/>
     </g>
     <path id="wire-3" role="graphics-symbol img" aria-label="wire connecting" aria-labelledby="wire-3 lightbulb-1 battery-1" class="wire" d="M180,60 V80 H20 V55"/>   
</svg>
```

Reply:
I don't think an assumption is the right thing to do here. That's not what assumptions are for. I like the idea of a theme, I would keep it as simple as possible. Maybe something mathematical? Like, "A rectangle is a closed shape with four right angle corners"?

Reply:
Yes "assumptions" have a precise sense in the ACT-Rules, and this is not that sense. It would rather be a note at the start of the examples Section saying that the svg used here are kept simple for the sake of not making the examples too big.

Note that even a simple yellow circle might convey information in some cases. We could imagine a weather forecast service that uses svg images to display the forecast and a yellow circle would mean "sunny weather". Adding a note in that sense at the start of the examples Section, and maybe also changing the titles and labels used in the examples to `"sunny"`, would be enough…

OTOH, most of the examples used in other rules are far from being "real" either… Having a single rule where the title attribute is not "This is a title" might be weird.

Reply:
We could simply add "How many corners are their in the yellow shape below?" making the examples part of a mathematical test and therefore making the images non decorative and making descriptions like "A yellow circle" important (although possibly reliant on colour alone... which is then an issue that should be fixed. Maybe dropping the "yellow" from both the description and the question would be enough).

Reply:
@ShadowBB, @WilcoFiers I do not agree that adding a question about the image will solve the issue.

It is certainly true that it makes it obviously "not decorative" and thus requiring accessible name and all that.

However, it completely changes the focus of the rule. Suddenly, `<svg>[yellow circle]</svg>` will pass (or be inapplicable) but `<p>How many circles?</p><svg>[Yellow circle]</svg>` will fail because in the first one the image is decorative and in the second one it isn’t.

That means, as a first step, that the target of this rule is **not** anymore the `<svg>` element. It is the `<svg>` element plus some surrounding context. So we have to cascade this change and figure out what the correct target is, which I foresee as being troublesome…

It also means that the rule cannot be automated since it now depends on some context, and on the actual meaning of the context. That is less of a problem (we are perfectly happy with having manual rules).

Lastly, SC 1.1.1 specifically lists non-text content that is a test/exercice as an exception. It is very easy to argue that having the accessible name being the answer to the question we're asking completely defeats the purpose of the exercice and thus falls within that exception…

---

All in all, I do not think that the rule should try and guess whether that specific element is decorative or not. If the designer/developer chose to make it visible, in the accessibility tree, and with a non-decorative role, then it should have an accessible name. Maybe it should not have been in the accessibility tree in the first place, and the fix is to add `aria-hidden=true`. Maybe it should have had a decorative role in the first place, and the fix is to add `role=presentation`. But if it's here in the tree, and with a non-decorative role, then it must have accessible name.

So, I do not know why a designer or developer would want to put a single yellow circle in the accessibility tree and with a non-decorative role. But I know that if they do they must add an accessible name to it.

The "decorative" exception on SC 1.1.1 says "If non-text content is pure decoration, (…) then it is implemented in a way that it can be ignored by assistive technology." I understand that as having `role=presentation` or `aria-hidden=true`, or something similar. If the yellow circle is pure decoration but is not implemented in a way that can be ignored (ie, is in the accessible tree with a non-decorative role), I think this does not satisfy SC 1.1.1 and it needs to be fixed.

---

I do not think it is a good idea to start discussing how much of the examples make sense "in real life". Almost none of them do (for any of the rules). We know they are just examples showing a very specific test case and are no way close to real web pages. I am afraid than trying to make the examples more "real"  will only result in obfuscating them by hiding their true purpose. The purpose of the examples here is not to check whether we are asking how many circles there are in the picture. The purpose of the examples here is to show that "visible + in the accessibility tree + non-decorative" implies "accessible name".

---

In short, I think the essence of many of the rules for SC 1.1.1 is "*If you put something in the accessibility tree*, either mark it as decorative or give it an accessible name". I do not think the rules should challenge **why** it is in the accessibility tree, or **why** it has a non-decorative role. I think the rules should only assess that it is in the accessibility tree with non-decorative role and perform checks accordingly.

Reply:
Hey Jym77, thanks for the long reply.

We are not asking to change these examples to make them pass/fail this rule. The context doesn't matter in determining whether or not is would pass/fail the rule. 

We simply ask for this context to be added so they passed examples would pass the entirety of WCAG (because we don't want passed examples that would fail other rules).

I agree with you that the answer to the question should not be in the example. Because that would indeed defeat the purpose of the test and thus trigger the SC 1.1.1 exception.

That is why I suggested "How many corners does this shape have?" with an accessible name of "1 circle"  because that would not give away the answer to the question.

Reply:
@ShadowBB 
> We simply ask for this context to be added so they passed examples would pass the entirety of WCAG (because we don't want passed examples that would fail other rules).

I'm afraid this is not the case. An example like [Passed example 1](https://act-rules.github.io/testcases/c4a8a4/3040ba5bd3485b7328c1cea8b5909982a10e4089.html) of the [Document title is descriptive](https://act-rules.github.io/rules/c4a8a4) rule is failing the [HTML has lang attribute](https://act-rules.github.io/rules/b5c3f8) rule and thus does not satisfy SC 3.1.1 "The default human language of each Web page can be programmatically determined. (Level A)"

And I don't think that is an isolated case.

Reply:
> 
> 
> @ShadowBB
> 
> > We simply ask for this context to be added so they passed examples would pass the entirety of WCAG (because we don't want passed examples that would fail other rules).
> 
> I'm afraid this is not the case. An example like [Passed example 1](https://act-rules.github.io/testcases/c4a8a4/3040ba5bd3485b7328c1cea8b5909982a10e4089.html) of the [Document title is descriptive](https://act-rules.github.io/rules/c4a8a4) rule is failing the [HTML has lang attribute](https://act-rules.github.io/rules/b5c3f8) rule and thus does not satisfy SC 3.1.1 "The default human language of each Web page can be programmatically determined. (Level A)"
> 
> And I don't think that is an isolated case.

Yes, this has not been the policy for long. I am sure we can find many passed examples that would fail WCAG. I think they should all be fixed.